addons/mail/models/ir_translation.py

```diff
@@ -0,0 +1,41 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models, _
+from odoo.exceptions import AccessError
+
+
+class IrTranslation(models.Model):
+    _inherit = 'ir.translation'
+
+
+    def create(self, vals_list):
+        res = super().create(vals_list)
+        res._check_is_dynamic()
+        return res
+
+    def write(self, vals):
+        res = super().write(vals)
+        self._check_is_dynamic()
+        return res
+
+    def _check_is_dynamic(self):
+        # if we don't modify translation of at least a model that inherits from mail.render.mixin, we ignore it
+        translations_for_mail_render_mixin = self.filtered(lambda tmpl: isinstance(self.env[tmpl.name.split(',')[0]], self.pool['mail.render.mixin']))
+        if not translations_for_mail_render_mixin:
+            return
+
+        # if we are admin, or that we can update mail.template we ignore
+        if self.env.is_admin() or self.env.user.has_group('mail.group_mail_template_editor'):
+            return
+
+        # Check that we don't add qweb code in translation when you don't have the rights
+        for trans in translations_for_mail_render_mixin:
+            res_model, res_id = trans.name.split(',')[0], trans.res_id
+            rec = self.env[res_model].with_context(lang=trans.lang).browse(res_id)
```

Comment:
Could you batch by model / res_ids ? Otherwise you are doing a single-record based computation in a method that is intended to be done in batch. This implies a bit of preparation but in the end it is better to batch when possible.